numpy/core/numeric.py

```diff
@@ -1383,22 +1387,18 @@ def roll(a, shift, axis=None):
     """
     a = asanyarray(a)
     if axis is None:
-        n = a.size
-        reshape = True
+        return roll(a.ravel(), shift, 0).reshape(a.shape)
     else:
-        try:
-            n = a.shape[axis]
-        except IndexError:
-            raise ValueError('axis must be >= 0 and < %d' % a.ndim)
-        reshape = False
-    if n == 0:
-        return a
-    shift %= n
-    indexes = concatenate((arange(n - shift, n), arange(n - shift)))
-    res = a.take(indexes, axis)
-    if reshape:
-        res = res.reshape(a.shape)
-    return res
+        indices = [arange(n).reshape([n] + [1] * ((-ax - 1) % a.ndim))
+                   for ax, n in enumerate(a.shape)]
+        for sh, ax in broadcast(shift, axis):
+            try:
+                indices[ax] -= sh
+            except IndexError:
+                raise ValueError("'axis' entry is out of bounds")
+        for index in indices:
+            index %= len(index)
```

Comment:
Could also do `-= sh % len(indices[ax])` and use negative indices. Doesn't really matter, bu `%` is not the fastest operation, so it might be faster. Could very well be a worthless micro optimization, though.